core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java

```diff
@@ -0,0 +1,261 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.avro;
+
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+import org.apache.avro.SchemaNormalization;
+import org.apache.avro.generic.GenericData;
+import org.apache.iceberg.AssertHelpers;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.mapping.MappedField;
+import org.apache.iceberg.mapping.MappedFields;
+import org.apache.iceberg.mapping.MappingUtil;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestAvroNameMapping extends TestAvroReadProjection {
+  @Override
+  protected GenericData.Record writeAndRead(String desc,
+                                            Schema writeSchema,
+                                            Schema readSchema,
+                                            GenericData.Record inputRecord) throws IOException {
+
+    // Use all existing TestAvroReadProjection tests to verify that
+    // we get the same projected (Avro) read schema whether we use
+    // NameMapping together with file schema without field-ids or we
+    // use a file schema having field-ids
+    GenericData.Record record = super.writeAndRead(desc, writeSchema, readSchema, inputRecord);
+    org.apache.avro.Schema expected = record.getSchema();
+    org.apache.avro.Schema actual = projectWithNameMapping(writeSchema, readSchema, MappingUtil.create(writeSchema));
+    Assert.assertEquals(expected, actual);
+    return record;
+  }
+
+  @Test
+  public void testNameMappingProjections() {
+    Schema fileSchema = new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(5, "locations", Types.MapType.ofOptional(6, 7,
+            Types.StringType.get(),
+            Types.StructType.of(
+                Types.NestedField.required(1, "lat", Types.FloatType.get()),
+                Types.NestedField.required(2, "long", Types.FloatType.get())
+            )
+        )));
+
+    // Table mapping does not project `locations` map
+    NameMapping nameMapping = MappingUtil.create(new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get())));
+
+    // Table read schema projects `locations` map's value
+    Schema readSchema = new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(5, "locations", Types.MapType.ofOptional(6, 7,
+            Types.StringType.get(),
+            Types.StructType.of(
+                Types.NestedField.required(2, "long", Types.FloatType.get())
+            )
+        )));
+
+    check(fileSchema, readSchema, nameMapping);
```

Comment:
I think that the assertions in the `check` method are correct, but it doesn't test what data will actually be read.

This test case, for example, will project null values for all of the locations maps. Checking that the actual schema and the readSchema have the same shape doesn't validate that you get null values when you read from Avro.

I think that the solution is to rewrite writeAndRead to actually write an Avro file without the IDs and then validate that the record matches. In this case, that `locations` is null.

Reply:
> This test case, for example, will project null values for all of the locations map

...

> In this case, that locations is null.

Doesn't it depend upon the file schema? The `locations` field is optional in the file schema. If the file has data for the `locations` column it will be returned, since the read/projected schema has the column projected.

Reply:
Here's the read schema which gets generated when using `NameMapping`
```
{
  "type" : "record",
  "name" : "table",
  "fields" : [ {
    "name" : "id",
    "type" : "long",
    "field-id" : 0
  }, {
    "name" : "locations",
    "type" : [ "null", {
      "type" : "map",
      "values" : [ "null", {
        "type" : "record",
        "name" : "r7",
        "fields" : [ {
          "name" : "long",
          "type" : "float",
          "field-id" : 2
        } ]
      } ],
      "key-id" : 6,
      "value-id" : 7
    } ],
    "default" : null,
    "field-id" : 5
  } ]
}
```
So locations will be filled if the input record has this field.

Reply:
The "locations" column should not be present. There was no field ID in the file schema and no field ID in the name mapping. That means the field cannot be projected and should be removed in column pruning.

Reply:
`PruneColumn` correctly removes it. The output of `AvroSchemaUtil.pruneColumns` is this
```
{
  "type" : "record",
  "name" : "table",
  "fields" : [ {
    "name" : "id",
    "type" : "long",
    "field-id" : 0
  } ]
}
```

But the `locations` map is added back by `BuildAvroProjection` .

Reply:
Since the `locations` map is part of the read schema, and if the file has a record where `locations` map is not null, even though the file schema has no id associated with the column, how can we null out the column? It seems we'd require a record level data transformation, no?

Reply:
Thanks! I thought you were saying that the schema with "locations" was produced by pruning. My mistake. It looks like this is a problem with how the new pruning interacts with `BuildAvroProjection`.

`BuildAvroProjection` assumes that if a column is missing from the file schema that is passed in, it is missing from the actual file schema, but in this case it isn't. To build the projection, it adds the missing field with the expectation that the resolving Avro decoder won't read data for that column and it will be filled in with nulls. That's not going to happen because the resolving decoder will match by name.

I think all we need to do is catch this case in `BuildAvroProjection` and change the field name to something that won't match in the actual file schema. Then it will have nulls filled in automatically.

I think this is a good demonstration of why we need to test these cases with a full round-trip to a data file and back.

Reply:
Hmm, let me think through this.

I think you are suggesting to change the read schema to not have the field (`locations`) if that field is not projected by NameMapping. For example, change the name to something different like `locations_[unique_suffix]` .  If that is the case, then we are effectively changing the projected/read schema. Won't this affect Spark, Presto etc, since they are expecting athe `locations` column? For instance, during query compile time the field `locations` was present as part of the read schema, but during execution or data reading it is no longer here. Do you think this is an issue?

Reply:
Another point I was thinking was -- if we maintain the invariant that the read schema is always a common subset of the table schema and table mapping, then we can side step this problem.

Happy to discuss this offline if this discussion is getting too lengthy.